### PR TITLE
[Autoflow] Rewrite Autoflow Tabular using new APIs

### PR DESCRIPTION
**Pull request by VWoeltjen**
Work-in-progress for #1810. Functionality is present, needs tests, documentation, and code cleanup. Opening PR early to allow early review.

TODO:

- [x] Separate out [row controller](https://github.com/nasa/openmct/pull/1816/files#r153015544)
- [x] Support [mutation](https://github.com/nasa/openmct/pull/1816/files#r153017408) (in-progress, currently gets stuck in loop)
- [x] Request [LAD data](https://github.com/nasa/openmct/pull/1816#pullrequestreview-79305103)
- [x] [Handle filtering in view](https://github.com/nasa/openmct/pull/1816/files#r153017114)
- [x] Move autoflowing of rows out of controller
- [x] Test autoflow behavior (changes to number of rows)
- [x] Test response to mutation
- [x] JSDoc

### CODE REVIEW

File: src/plugins/autoflow/AutoflowTabularController.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    AutoflowTabularController.prototype.updateRow = function (row, formatRange, formatDomain, evaluate, datum) {
+        row.value = formatRange(datum);
+        row.classes = evaluate(datum).map(function (violation) {
+            return violation.cssClass;
+        }).join(" ");
+
+        this.data.updated = formatDomain(datum);
+    };
+
+    AutoflowTabularController.prototype.makeRow = function (childObject) {
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularController.js]: Move row behavior to a AutoflowTabularRowController?

File: src/plugins/autoflow/AutoflowTabularController.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    };
+
+    AutoflowTabularController.prototype.activate = function () {
+        if (this.active) {
+            this.destroy();
+        }
+        this.active = true;
+
+        this.openmct.composition.get(this.domainObject)
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularController.js]: For better mutation support (without having to listen on mutation), you can do

```
this.composition = this.openmct.composition.get(this.domainObject);
this.composition.on('add', this.addRow);
this.composition.on('remove', this.removeRow);
this.composition.load();
```

File: src/plugins/autoflow/AutoflowTabularController.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        var changed = this.rowCount !== rows;
+        this.rowCount = rows;
+        if (changed) {
+            this.update();
+        }
+    };
+
+    AutoflowTabularController.prototype.activate = function () {
+        if (this.active) {
+            this.destroy();
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularController.js]: What exercises / requires this codepath?  It doesn't make sense to me that a component with a lifecycle can be reincarnated for a second round.

File: src/plugins/autoflow/AutoflowTabularController.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+        this.rows[id] = row;
+
+        return row;
+    };
+
+    AutoflowTabularController.prototype.update = function () {
+        var column = [];
+        var index = 0;
+        var filteredObjects =
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularController.js]: Filtering could be a view-only concern?

File: src/plugins/autoflow/AutoflowTabularController.js [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    AutoflowTabularController.prototype.setRows = function (rows) {
+        var changed = this.rowCount !== rows;
+        this.rowCount = rows;
+        if (changed) {
+            this.update();
+        }
+    };
+
+    AutoflowTabularController.prototype.activate = function () {
+        this.destroy();
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularController.js]: What exercises / requires this codepath? It doesn't make sense to me that a component with a lifecycle can be reincarnated for a second round.

File: src/plugins/autoflow/AutoflowTabularController.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        this.active = true;
+
+        this.openmct.composition.get(this.domainObject)
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularController.js]: For better mutation support in tandem with per-row controllers, you can do something like:

```
this.composition = this.openmct.composition.get(this.domainObject);
this.composition.on('add', this.addRow, this);
this.composition.on('remove', this.removeRow, this);
this.composition.load();
```

File: src/plugins/autoflow/AutoflowTabularController.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        var ranges = metadata.valuesForHints(['range']);
+        var domains = metadata.valuesForHints(['domain']);
+        var rangeFormatter =
+            this.openmct.telemetry.getValueFormatter(ranges[0]);
+        var domainFormatter =
+            this.openmct.telemetry.getValueFormatter(domains[0]);
+        var evaluator =
+            this.openmct.telemetry.limitEvaluator(childObject);
+
+        this.unlistens.push(this.openmct.telemetry.subscribe(
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularController.js]: Should perform a LAD request for telemetry as well.

**Comment by VWoeltjen** [comment]
### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

**Comment by VWoeltjen** [comment]
@larkin Ready for another look! Assuming this will be squash-merged or similar, happy to rebase if that is preferred.

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

File: src/plugins/autoflow/AutoflowTabularPluginSpec.js [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+                var mockComposition;
+                var mockMetadata;
+                var mockEvaluator;
+                var mockUnsubscribes;
+                var callbacks;
+                var view;
+
+                function waitsForChange() {
+                    var oldHtml = testContainer.innerHTML;
+                    waitsFor(function () {
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularPluginSpec.js]: Would it make sense to use `Vue.nextTick()` per [their testing docs](https://vuejs.org/v2/guide/unit-testing.html#Asserting-Asynchronous-Updates) for waiting for updates?

File: src/plugins/autoflow/AutoflowTabularRowController.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        this.unsubscribe = this.openmct.telemetry.subscribe(
+            this.domainObject,
+            this.updateRowData.bind(this)
+        );
+
+        this.openmct.telemetry.request(
+            this.domainObject,
+            { size: 1 }
+        ).then(function (history) {
+            if (!this.initialized && history.length > 0) {
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularRowController.js]: This appears to be preventing a race condition; I guess it's for "I got realtime data before LAD response?"

File: src/plugins/autoflow/AutoflowTabularRowController.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        this.data.classes = violations ? violations.cssClass : "";
+        this.data.value = this.rangeFormatter.format(datum);
+        this.callback(this.domainFormatter.format(datum));
+    };
+
+    /**
+     * Activate this controller; begin listening for changes.
+     */
+    AutoflowTabularRowController.prototype.activate = function () {
+        this.destroy();
```
Comment (larkin) [inline src/plugins/autoflow/AutoflowTabularRowController.js]: This just feels weird to me-- destroying at the beginning of activate?

In looking at it, I don't think activate is ever called twice, is it?

Comment (VWoeltjen) [reply src/plugins/autoflow/AutoflowTabularRowController.js]: You're correct, this is a remnant of unnecessary lifecycle guards that I didn't completely remove

File: src/plugins/autoflow/AutoflowTabularPluginSpec.js (hunk 8 above)
Comment (VWoeltjen) [reply src/plugins/autoflow/AutoflowTabularPluginSpec.js]: Thanks for the link! I'll try that out

File: src/plugins/autoflow/AutoflowTabularRowController.js (hunk 9 above)
Comment (VWoeltjen) [reply src/plugins/autoflow/AutoflowTabularRowController.js]: That's correct; don't want to clobber realtime data with historical data

File: src/plugins/autoflow/AutoflowTabularPluginSpec.js (hunk 8 above)
Comment (VWoeltjen) [reply src/plugins/autoflow/AutoflowTabularPluginSpec.js]: That works well, but I opted for `window.requestAnimationFrame`, which accomplishes the same in practice, but allows the spec to remain totally Vue-agnostic

File: src/plugins/autoflow/AutoflowTabularRowController.js (hunk 9 above)
Comment (larkin) [reply src/plugins/autoflow/AutoflowTabularRowController.js]: Is this worthy of a test case?

**Pull request merged**